src/compiler/checker.ts

```diff
@@ -3690,6 +3697,48 @@ module ts {
             }
             return links.resolvedType;
         }
+        
+        function getTypeGuardType(node: TypeGuardTypeNode): Type {
+            let type = <TypeGuardType>createType(TypeFlags.TypeGuard);
+            let hasParameters = false;
+            type.parameterIndex = -1;
+            if (node.target) {
+                let functionDeclaration = <SignatureDeclaration>node.target.parent.parent;

```

Comment:
nothing says that this is a function declaration. you should be checking this earlier, and giving an error if a typeGuardType is used anywhere else.